pkg/server/ring.go

```diff
@@ -0,0 +1,275 @@
+	delegate = ring.NewLeaveOnStoppingDelegate(delegate, logger)
+	delegate = ring.NewAutoForgetDelegate(heartbeatTimeout*2, delegate, logger)
+
+	lifecycler, err := ring.NewBasicLifecycler(
+		lifecyclerCfg,
+		ringName,
+		ringKey,
+		ringStore,
+		delegate,
+		logger,
+		reg,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize ruler's lifecycler: %s", err)
+	}
+
+	rulerRing, err := ring.NewWithStoreClientAndStrategy(
+		toRingConfig(ms.cfg, KVStore),
+		ringName,
+		ringKey,
+		ringStore,
+		ring.NewIgnoreUnhealthyInstancesReplicationStrategy(),
+		reg,
+		logger,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize ruler's ring: %s", err)
+	}
+
+	startFn := func(ctx context.Context) error {
+		err := memberlistKVsvc.StartAsync(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to start memberlist service: %s", err)
+		}
+		err = rulerRing.StartAsync(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to start the ring: %s", err)
+		}
+		err = lifecycler.StartAsync(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to start the lifecycler: %s", err)
+		}
+		err = pool.StartAsync(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to start the ring client pool: %s", err)
+		}
+
+		logger.Info("waiting until resource server is JOINING in the ring")
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+		defer cancel()
+		if err := ring.WaitInstanceState(ctx, rulerRing, lifecycler.GetInstanceID(), ring.JOINING); err != nil {
+			return fmt.Errorf("error switching to JOINING in the ring: %s", err)
+		}
+		logger.Info("resource server is JOINING in the ring")
+
+		if err := lifecycler.ChangeState(context.Background(), ring.ACTIVE); err != nil {
+			return fmt.Errorf("error switching to ACTIVE in the ring: %s", err)
+		}
+
+		logger.Info("waiting until resource server is ACTIVE in the ring")
+		if err := ring.WaitInstanceState(context.Background(), rulerRing, lifecycler.GetInstanceID(), ring.ACTIVE); err != nil {
+			return fmt.Errorf("error switching to ACTIVE in the ring: %s", err)
+		}
+		logger.Info("resource server is ACTIVE in the ring")
+
+		return nil
+	}
+
+	errChan := make(chan error)
+	runningFn := func(ctx context.Context) error {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-errChan:
+			return err
+		}
+	}
+
+	stoppingFn := func(failureReason error) error {
+		return nil
+	}
+
+	ms.distributor = &resource.Distributor{
+		ClientPool: pool,
+		Ring:       rulerRing,
+		Lifecycler: lifecycler,
+	}
+
+	ms.httpServerRouter.Path("/ring").Methods("GET", "POST").Handler(rulerRing)
+
+	svc := services.NewBasicService(startFn, runningFn, stoppingFn)
+
+	return svc, nil
+}
+
+func toMemberlistConfig(cfg *setting.Cfg) *memberlist.KVConfig {
+	memberlistKVcfg := &memberlist.KVConfig{}
+	flagext.DefaultValues(memberlistKVcfg)
+	memberlistKVcfg.MetricsNamespace = ringName
```

Comment:
Memberlist KV store is typically shared between all rings used by the instance, so using single ring's name as metrics namespace seems weird. Perhaps we could keep it empty or use something generic like `grafana`.